### DISCUSSION TITLE
Macro compilation issue in modularized project

### DISCUSSION THREAD

**Discussion by gyulavoros** [pointfreeco/swift-composable-architecture#3292] [General]
### Description

I'm looking for guidance on how to integrate TCA with multiple feature modules.

Given that multiple framework targets depend on TCA, project compilation breaks with the following Xcode error:

```
External macro implementation type 'PerceptionMacros.PerceptibleMacro' could not be found for macro 'Perceptible()'
```

**Project setup:**
- Vanilla Xcode project, or `xcodeproj` generated via [XcodeGen](github.com/yonaskolb/xcodegen)
- TCA is added as SPM dependency
- Two framework targets + one app target
- **Both framework targets depend on TCA** (this breaks macro compilation)
- App target depends on both framework targets

```
|-- App
|---- FeatureA
|------ TCA
|---- FeatureB
|------ TCA
```

This setup used to work with older TCA versions (`1.11.2` and prior). I have limited knowledge of macro internals and having a hard time to pinpoint what the issue is.

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The project compiles.

### Actual behavior

Compilation fails with:

```
External macro implementation type 'PerceptionMacros.PerceptibleMacro' could not be found for macro 'Perceptible()'
```

![CleanShot 2024-08-19 at 20 06 53@2x](https://github.com/user-attachments/assets/8c0692a4-7004-4395-9b0e-85e756ec75bd)

### Steps to reproduce

I attached a minimal example demonstrating the issue:
[ModularizedMacroIssue.zip](https://github.com/user-attachments/files/16663781/ModularizedMacroIssue.zip)

### The Composable Architecture version information

1.13.0

### Destination operating system

iOS 17.5

### Xcode version information

Xcode 15.4 (15F31d)

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @gyulavoros, this is not an issue with the library per se, and instead most likely an issue with how your project is set up in Xcode. I am going to convert this to a discussion and perhaps someone can help you out over there.

**Reply by gyulavoros** [reply]
Thank you, @mbrandonw, for taking a look!

To flip the question, are you aware of any extra setup necessary in a new Xcode project, beyond adding `ComposableArchitecture` in the `Frameworks and Libraries` menu to make it work, given multiple framework targets?

![CleanShot 2024-08-19 at 20 47 25@2x](https://github.com/user-attachments/assets/a9d30d15-66e9-48ef-a9f8-06713cb12d76)

**Reply by gyulavoros** [reply]
To confirm, that downgrading `ComposableArchitecture` to version `1.11.2` makes the project to compile just fine.

![CleanShot 2024-08-20 at 14 07 04@2x](https://github.com/user-attachments/assets/b9e0302d-2749-4f09-a018-0f5aa41a9c93)

![CleanShot 2024-08-20 at 14 07 13@2x](https://github.com/user-attachments/assets/c5e2cdad-4603-44cc-ab4d-1c3bbc496992)

As the compile error comes from `swift-navigation`, I'm wondering if this could be related to the `swiftui-navigation` → `swift-navigation` migration/transition?

![CleanShot 2024-08-20 at 14 11 33@2x](https://github.com/user-attachments/assets/b1a27e50-5109-4ffb-8937-8a69a404d267)

@mbrandonw if you have any two cents or insights on the issue, it would be super helpful!

**Reply by mbrandonw** [reply]
Hi @gyulavoros, it does not compile for me even if I switch TCA to 1.11:

<img width="398" alt="image" src="https://github.com/user-attachments/assets/0bfe1af8-9f84-439c-b980-3749e1bbc488">

The problem is that you have two dynamic frameworks depending on TCA. You can't do this. What you need to do is create a 3rd common framework that depends on TCA:

<img width="836" alt="image" src="https://github.com/user-attachments/assets/fa7a7c6b-14da-41de-a6a9-0d308f6a34f8">

And then have each of FeatureA and FeatureB depend on the Common framework:

<img width="909" alt="image" src="https://github.com/user-attachments/assets/5c919e33-b674-4fd5-bd92-4564bd7b1c2d">

But honestly I don't think it's worth dealing with Xcode frameworks. You should just create an SPM package that houses FeatureA and FeatureB, and then they are free to depend on TCA.

**Reply by gyulavoros** [reply]
Bingo, it works with a shared framework target, thank you so much for the tip! 🙏 @mbrandonw 

Not being able to depend on `ComposableArchitecture` from multiple dynamic frameworks is due to TCA being statically linked, or it is specifically something that you can't do when macros are involved?

As for Xcode framework targets, we use Xcodegen to configure many other aspects of our project (dependencies, modules, schemes, info.plists, build plugins, and various Xcode build settings), so migrating to a `Package.swift` file based setup would not be trivial.

**Reply by mbrandonw** [reply]
> Not being able to depend on `ComposableArchitecture` from multiple dynamic frameworks is due to TCA being statically linked, or it is specifically something that you can't do when macros are involved?

It is due to depending on a static library from dynamic libraries, and is just a general limitation of the build system.

**Reply by gyulavoros** [reply]
In case anyone bumps into this discussion, I have one more (unfortunate) update: although linking TCA to only one "common" framework enables the project to compile, I faced two more issues:
- I cannot build the project via CLI (e.g, `xcodebuild build -project <project> -scheme <scheme> -sdk iphonesimulator | xcbeautify`) because of the same macro error. Only building from Xcode works.
- Xcode is acting out almost all TCA related code are marked with errors and code completion doesn't work. Again, while Build and Run works.

![CleanShot 2024-08-21 at 16 09 34@2x](https://github.com/user-attachments/assets/9dc7cb9a-312a-4e9a-a672-3bbfb20cf465)

Eventually, I'm following Brandon's advice and drop Xcode framework targets and instead migrate to SPM package libraries. I was able to combine Xcodegen and SPM local packages in a way that feels reasonable. Modules are defined in a `Package.swift` file and the application, extensions, schemes and a few other things are still configured with Xcodegen.

**Reply by iranjith4** [reply]
I faced the exact issue in XCode 16.1 Beta. We have a common package that has the TCA dependency. Before running the main target, build the common package with the TCA once. After building the Common package, run the app; these errors should disappear.

**Reply by gyulavoros** [reply]
I might be wrong here, as it was a while ago when I was wrangling with this, but I remember that custom Build Configurations could also interfere with macros. In SPM, "Debug" and "Release" are baked in, and the only available configurations. If your application target scheme uses custom build configurations (e.g, "Mock", "Staging", etc.) Xcode breaks and it displays the errors above. Previews don't work either.

What's frustrating, is that you can still compile, build, and run the project on both the simulator or a physical device. However, not having code completion and tripping over hundreds of inline errors is a dealbreaker.

To sum up, keeping the project setup "as vanilla as possible", defining modules as SPM packages, relying on the default "Debug" and "Release" configurations, etc. is the way to go, I can confirm that both Composable Architecture and other macro based dependencies work like a charm.

**Reply by rhkrthdud7** [reply]
@gyulavoros Hey I just renamed my configuration as Debug / Release and all the macro errors disappeared.
I was struggling for weeks because of this issue, code completion was completely broken.
But you just saved my life thank you. 👍

**Reply by shmoylove** [reply]
I want to share my experience of solving the problem. I updated Xcode to 16.1 (16B40) and the problem was solved. I also use **SPM** and install the dependency inside the package file

**Reply by adarhef** [reply]
I want to add to this as well.
I use Tuist and I link all my dependencies dynamically. Partly this is because I need Previews to work, and they just don't with Tuist unless dependencies are linked as frameworks.
Everything works fine for the most part, but yesterday I realized Archiving is broken with this same macro error.
Release builds are fine, I can Profile without a problem for example.
I don't know what's special about Archiving for it to act up like this. Build logs are no help.
Would appreciate any pointers

**Reply by mbrandonw** [reply]
Hi @adarhef, this sounds more like a Tuist problem and so I recommend you start a discussion on that repo to see if they have any suggestions.

**Comment by protonjohn** [comment]
It's also worth noting that `Sources/SwiftNavigation/UIBinding.swift` in the `swift-navigation` project does not `import Perception` despite using the `@Perceptible` macro for `_UIBindingWrapper`, which might be in some way related to Xcode's unhappiness. (It didn't seem to help when I fixed that, though.)

**Reply by mbrandonw** [reply]
Hi @protonjohn, SwiftNavigation does an `@_exported import` of Perception, so it should be available to every file in the library.
